### ISSUE TITLE
curl failed to parse AWS Gateway API response 

### ISSUE THREAD

**Issue by AzlinII** [jeroen/curl#223]
I'm running into an issue where I'm getting a response back from an AWS Gateway API but curl seems to be failing to parse the response out at the last step and is throwing an error. After digging into it a bit, I think it is due to the libcurl version being the issue. I have two environments where I run the exact same code with the exact same package version to call the API but the environment that had libcurl version 7.58.0 fails while the environment that had libcurl version 7.47.0 works.  

<details>
<summary>Here is the code I have.  It uses the httr package but the error it throws shows it's an issue with curl.</summary>

```
aws_get <- function(url, aws_key = Sys.getenv("AWS_ACCESS_KEY_ID"), 
                    aws_secret = Sys.getenv("AWS_SECRET_ACCESS_KEY"), 
                    region = "us-east-1",
                    extra_headers = NULL, 
                    query_args = list()){
  url <- URLencode(url)
  url_parts <- httr::parse_url(url)
  host <- url_parts$hostname
  action <- paste0("/", url_parts$path)
  
  amz_date <- format(Sys.time(), format = "%Y%m%dT%H%M%SZ")
  canonical_headers <- c("Host" = host, "X-Amz-Date" = amz_date,
                         extra_headers)
  signature_info <- aws.signature::signature_v4_auth(
    datetime = amz_date,
    region = region,
    verb = "GET",
    query_args = query_args,
    action = action,
    request_body = "",
    service = "execute-api",
    secret = aws_secret,
    key = aws_key,
    canonical_headers = canonical_headers
  )
  headers <- httr::add_headers(Authorization = signature_info$SignatureHeader,
                               .headers = canonical_headers)
  httr::GET(url, headers, httr::verbose())
}
```
</details>

<details>
  <summary>Here is the response and curl_version info I get in the environment that failed.</summary>

![github_issue_img](https://user-images.githubusercontent.com/15007726/82471415-47449f80-9a8c-11ea-8367-77d57c16ffdf.png)

![image](https://user-images.githubusercontent.com/15007726/80844151-a5880c00-8bcb-11ea-8506-661e2ffffbe3.png)
</details>

<details>
<summary>Here is the response and curl_version info I get in the environment that worked.</summary>

![github_issue_img2](https://user-images.githubusercontent.com/15007726/82471478-5e838d00-9a8c-11ea-8947-6945000be990.png)

![image](https://user-images.githubusercontent.com/15007726/80844169-b769af00-8bcb-11ea-8859-37c3698050cd.png)

</details>

As you can see from the verbose, the response return is almost the same with the exception of the User-Agent libcurl version.

**Comment by jeroen** [comment]
Dupe of https://github.com/jeroen/curl/issues/179 and https://github.com/jeroen/curl/issues/160.

**Comment by AzlinII** [comment]
@jeroen I have read both of those but they don't really solve my problem. You answered in those that it's due to the Content-Encoding not being part of the http spec, but as you can see from the request that worked, it has the same Content-Encoding as the one that failed.

The only difference I could find between the two is the libcurl version.  Now if the http spec changed between the two version of libcurl there is nothing that can be done about it that but outside of that, I don't know why it would work with one version of libcurl but not the other. 

I may be overlooking something though so if you see something or know of something that could be different between the two that could have lead to one working while the other not, please let me know.

**Comment by AzlinII** [comment]
@jeroen I have looked into this some more and I believe the thing causing the issue is partially due to this package. I have tried to make the call in the same environment to using both python and libcurl directly. Both of those calls were a success. 

After comparing the python code to the R code, I see that in python, it doesn't try to decode the response as soon as it is return from the API call.
```
import http.client
import mimetypes
conn = http.client.HTTPSConnection("xxxxxxxxxxx.execute-api.us-east-1.amazonaws.com")
payload = ""
headers = {
  "X-Amz-Date": "20200520T155133Z",
  "Authorization": "AWS4-HMAC-SHA256 Credential=xxxxxxxxxxxxxxxxx/20200520/us-east-1/execute-api/aws4_request, SignedHeaders=host;x-amz-date, Signature=xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx"
}
conn.request("GET", "/Dev/projects/SEQ39333", payload, headers)
res = conn.getresponse()
data = res.read()
print(data.decode("utf-8"))
```
while in R, or at least according to the error message, is trying to decode it once it returns but seems to be failing because libcurl could not decode it.  

I am not sure not sure if there is a way to prevent libcurl from trying to decode the response right away in your package but if there is, please let me know.  

Here is the format of the libcurl call that worked
```
curl --location --request GET "https://xxxxxxxxxxx.execute-api.us-east-1.amazonaws.com/Dev/projects/SEQ39333" --header "X-Amz-Date: 20200520T164115Z" --header "Authorization: AWS4-HMAC-SHA256 Credential=xxxxxxxxxxxxxxxxx/20200520/us-east-1/execute-api/aws4_request, SignedHeaders=host;x-amz-date, Signature=xxxxxxxxxxxxxxxxxxxx"
```

**Comment by AzlinII** [comment]
@jeroen It's been a while but I think I have finally found a solution to this issue. As I kind of expected, it was an issue with libcurl decoding the content return from the API call when it didn't need to be decoded. I ended up setting `http_content_decode = 0` and was able to solve my error of 
```
 Error in curl::curl_fetch_memory(url, handle = handle) : 
  Unrecognized content encoding type. libcurl understands deflate, gzip content encodings. 
```
I can't say for certain but I think maybe with the more recent version of libcurl, it automatically set `CURLOPT_HTTP_CONTENT_DECODING ` to true while in the older version it didn't. That was why with the older version of libcurl, your package was still able to work without issues. 

Anyway, while both [#177](https://github.com/jeroen/curl/issues/179) and [#160](https://github.com/jeroen/curl/issues/160) are closed, it might be good to let them try using `http_content_decode = 0` as well and see if that works. Below is the full example I used.

```
h <- curl::new_handle()
curl::handle_setopt(h, http_content_decoding = 0)
res <- curl::curl_fetch_memory(url, h)
```
